### PR TITLE
Add basic authentication

### PR DESCRIPTION
**Pull request by krabobobr**
### Jira-link:
https://virtocommerce.atlassian.net/browse/PT-9286

### CODE REVIEW

File: src/VirtoCommerce.Platform.Web/Startup.cs [hunk 1]
```diff
@@ -387,9 +389,9 @@ public void ConfigureServices(IServiceCollection services)
 
             services.AddAuthorization(options =>
             {
-                //We need this policy because it is a single way to implicitly use the two schema (JwtBearer and ApiKey)  authentication for resource based authorization.
+                //We need this policy because it is a single way to implicitly use the three schemas (JwtBearer, ApiKey and Basic) authentication for resource based authorization.
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Web/Startup.cs:392]: could you also  add a new schema here please https://github.com/VirtoCommerce/vc-platform/blob/dev/src/VirtoCommerce.Platform.Web/Security/Authorization/PermissionAuthorizationPolicyProvider.cs#L50

**Review by tatarincev — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.Platform.Web/Startup.cs (hunk 1 above)
Comment (krabobobr) [reply src/VirtoCommerce.Platform.Web/Startup.cs:392]: thanks for noticing that. added.

File: src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        private readonly UserManager<ApplicationUser> _userManager;
+        private readonly SignInManager<ApplicationUser> _signInManager;
+
+        public BasicAuthenticationHandler(
+            IOptionsMonitor<BasicAuthenticationOptions> options,
+            ILoggerFactory logger,
+            UrlEncoder encoder,
+            ISystemClock clock,
+            UserManager<ApplicationUser> userManager,
+            SignInManager<ApplicationUser> signInManager) : base(options, logger, encoder, clock)
```
Comment (artem-dudarev) [inline src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs:25]: Move `: base(options, logger, encoder, clock)` to a separate line

File: src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            var authHeader = Request.Headers["Authorization"].ToString();
+            if (authHeader != null && authHeader.StartsWith("basic", StringComparison.OrdinalIgnoreCase))
+            {
+                try
+                {
+                    var encodedAuth = authHeader.Substring("Basic ".Length).Trim();
```
Comment (artem-dudarev) [inline src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs:38]:
```suggestion
            const string scheme = "Basic ";
            var authHeader = Request.Headers["Authorization"].ToString();

            if (authHeader != null && authHeader.StartsWith(scheme, StringComparison.OrdinalIgnoreCase))
            {
                try
                {
                    var encodedCredentials = authHeader.Substring(scheme.Length).Trim();
```

File: src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+                    if (string.IsNullOrEmpty(encodedAuth))
+                    {
+                        return AuthenticateResult.Fail("Invalid authentication provided, access denied.");
+                    }
+
+                    var loginPass = DecodeUserIdAndPassword(encodedAuth);
```
Comment (artem-dudarev) [inline src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs:45]:
```suggestion
                    var (userName, password) = DecodeUserNameAndPassword(encodedCredentials);
```

File: src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+                    if (_userManager.SupportsUserLockout && await _userManager.IsLockedOutAsync(user))
+                    {
+                        return AuthenticateResult.Fail($"User { user.UserName } is currently locked out.");
+                    }
+
+                    if (!await _userManager.CheckPasswordAsync(user, loginPass.Password))
+                    {
+                        return AuthenticateResult.Fail("Invalid password provided, access denied.");
+                    }
```
Comment (artem-dudarev) [inline src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs:65]: We should not reveal such details. The message should be the same in all cases: Invalid user name or password.

File: src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+        private static (string Login, string Password) DecodeUserIdAndPassword(string encodedAuth)
+        {
+            var loginPass = Encoding.UTF8.GetString(Convert.FromBase64String(encodedAuth));
+
+            var separator = loginPass.IndexOf(':');
+            if (separator == -1) throw new InvalidOperationException("Invalid Authorization header: Missing separator character ':'. See RFC2617.");
+
+            return (loginPass.Substring(0, separator), loginPass.Substring(separator + 1));
+        }
```
Comment (artem-dudarev) [inline src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs:88]:
```suggestion
        private static (string userName, string password) DecodeUserNameAndPassword(string encodedCredentials)
        {
            var decodedCredentials = Encoding.UTF8.GetString(Convert.FromBase64String(encodedCredentials));

            var separatorPosition = decodedCredentials.IndexOf(':');
            if (separatorPosition == -1)
            {
                throw new InvalidOperationException("Invalid Authorization header value: Missing separator character ':'. See RFC2617.");
            }

            return (decodedCredentials.Substring(0, separatorPosition), decodedCredentials.Substring(separatorPosition + 1));
        }
```

File: src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                    }
+
+                    if (!await _userManager.CheckPasswordAsync(user, loginPass.Password))
+                    {
+                        return AuthenticateResult.Fail("Invalid password provided, access denied.");
+                    }
+
+                    var claimsPrincipal = await _signInManager.CreateUserPrincipalAsync(user);
+                    var ticket = new AuthenticationTicket(claimsPrincipal, Options.Scheme);
+                    return AuthenticateResult.Success(ticket);
```
Comment (artem-dudarev) [inline src/VirtoCommerce.Platform.Web/Security/Authentication/BasicAuthenticationHandler.cs:69]:
```suggestion
                    var ticket = new AuthenticationTicket(claimsPrincipal, Options.Scheme);

                    return AuthenticateResult.Success(ticket);
```

**Review by artem-dudarev — CHANGES_REQUESTED** [review]

**Review by artem-dudarev — APPROVED** [review]

**Review by tatarincev — APPROVED** [review]

**Pull request merged**